pkg/graveler/retention/expired_commits.go

```diff
@@ -12,37 +14,84 @@ type GarbageCollectionCommits struct {
+	Version      graveler.CommitVersion
+}
+
+func NewCommitNode(commitID graveler.CommitID, creationDate time.Time, parentsIDs []graveler.CommitID, version graveler.CommitVersion) CommitNode {
+	return CommitNode{
+		CommitID:     commitID,
+		CreationDate: creationDate,
+		ParentsIDs:   parentsIDs,
+		Version:      version,
+	}
+}
+
+var ErrCommitNotFound = errors.New("commit not found")
+
 // GetGarbageCollectionCommits returns the sets of expired and active commits, according to the repository's garbage collection rules.
 // See https://github.com/treeverse/lakeFS/issues/1932 for more details.
 // Upon completion, the given startingPointIterator is closed.
 func GetGarbageCollectionCommits(ctx context.Context, startingPointIterator *GCStartingPointIterator, commitGetter *RepositoryCommitGetter, rules *graveler.GarbageCollectionRules, previouslyExpired []graveler.CommitID) (*GarbageCollectionCommits, error) {
 	// From each starting point in the given startingPointIterator, it iterates through its main ancestry.
 	// All commits reached are added to the active set, until and including the first commit performed before the start of the retention period.
 	// All further commits in the ancestry are added to the expired set. The iteration stops upon reaching a commit which exists in the previouslyExpired set, or the DAG root.
-	now := time.Now()
+	var (
+		commitNode CommitNode
+		ok         bool
+	)
 	processed := make(map[graveler.CommitID]time.Time)
 	previouslyExpiredMap := make(map[graveler.CommitID]bool)
 	for _, commitID := range previouslyExpired {
 		previouslyExpiredMap[commitID] = true
 	}
 	activeMap := make(map[graveler.CommitID]struct{})
 	expiredMap := make(map[graveler.CommitID]struct{})
+
+	commitsIterator, err := commitGetter.ListCommits(ctx)
+	if err != nil {
+		return nil, err
+	}
+	isIncremental := len(previouslyExpired) > 0
+	commitsMap := make(map[graveler.CommitID]CommitNode)
+	if !isIncremental {
+		defer commitsIterator.Close()
+		for commitsIterator.Next() {
+			commitRecord := commitsIterator.Value()
+			commitsMap[commitRecord.CommitID] = NewCommitNode(commitRecord.CommitID, commitRecord.Commit.CreationDate, commitRecord.Commit.Parents, commitRecord.Commit.Version)
+		}
+	}
+
+	now := time.Now()
 	defer startingPointIterator.Close()
 	for startingPointIterator.Next() {
 		startingPoint := startingPointIterator.Value()
 		retentionDays := int(rules.DefaultRetentionDays)
-		commit, err := commitGetter.GetCommit(ctx, startingPoint.CommitID)
-		if err != nil {
-			return nil, err
+		if isIncremental {
+			var commit *graveler.Commit
+			commit, err = commitGetter.GetCommit(ctx, startingPoint.CommitID)
+			if err != nil {
+				return nil, err
+			}
+			commitNode = NewCommitNode(startingPoint.CommitID, commit.CreationDate, commit.Parents, commit.Version)
+		} else {
+			commitNode, ok = commitsMap[startingPoint.CommitID]
+			if !ok {
+				return nil, fmt.Errorf("%w: %s", ErrCommitNotFound, startingPoint.CommitID)
+			}
```

Comment:
If you abstract away the commits fetching mechanism inside `commitGetter`, then you wouldn't have to branch (`if isIncremental {...} else {...}`) every time you fetch a commit.